### ISSUE TITLE
[Type] using onSettled or onSuccess messes with the type inference

### ISSUE THREAD

**Issue by Ayc0** [tannerlinsley/react-query#730]
**Describe the bug**

Using `onSuccess` or `onSettled` messes with the inference and `data` becomes unknown.

I don't know if this is an issue with TS or with react-query but I think TS tries to infer response's type from the fetcher AND from the config object at the same time. And as it fails, it falls back to `unknown`.

**To Reproduce**

I made a code sandbox with this issue:
https://codesandbox.io/s/react-query-error-jvseh?file=/src/App.tsx

**Expected behavior**

`response`'s type in both onSuccess and onSettled should be auto-inferred

**Screenshots**

![image](https://user-images.githubusercontent.com/22725671/87145157-30127900-c2a9-11ea-927c-2415b8ae2fcf.png)

**Additional informations:**
 - react-query: 2.4.14
 - typescript: 3.9.3 (I also tried on 3.9.6 and it was the same thing)

**Comment by Bene8493** [comment]
I had the same problem. As a workaround you can use the object syntax:
```javascript
const {data} = useQuery({
        queryKey: ["key1", "key2"],
        queryFn: foo,
        config: {
            onSuccess: (response) => console.log(response),
        }
    })
```

**Comment by jansedlon** [comment]
Same issue :)

**Comment by franjs** [comment]
As a workaround you could also either pass the type arguments to `useQuery` (TResult and TKey):

```
const { data } = useQuery<string, ["key1", "key2"]>(["key1", "key2"], foo, {
    onSuccess: response => console.log(response),
});
```

or add the parameters type to onSuccess/onSettled (TResult):

```
const { data } = useQuery(["key1", "key2"], foo, {
    onSuccess: (response: string) => console.log(response),
});
```

I fixed the issue by changing the type definitions from:

```
export type UseQueryRest<TResult, TKey extends AnyQueryKey, TError> =
  | []
  | [QueryFunction<TResult, TKey>]
  | [QueryFunction<TResult, TKey>, QueryOptions<TResult, TError>]
  | [QueryOptions<TResult, TError>]

export function useQuery<TResult, TKey extends AnyQueryKey, TError = Error>(
  queryKey: TKey | false | null | undefined,
  ...rest: UseQueryRest<TResult, TKey, TError>
): QueryResult<TResult, TError>

export function useQuery<TResult, TKey extends string, TError = Error>(
  queryKey: TKey | false | null | undefined,
  ...rest: UseQueryRest<TResult, [TKey], TError>
): QueryResult<TResult, TError>

```

to:

```
export function useQuery<TResult, TKey extends AnyQueryKey, TError = Error>(
  queryKey: TKey | false | null | undefined,
  queryFn?: QueryFunction<TResult, TKey>,
  queryOptions?: QueryOptions<TResult, TError>
): QueryResult<TResult, TError>

export function useQuery<TResult, TKey extends string, TError = Error>(
  queryKey: TKey | false | null | undefined,
  queryFn?: QueryFunction<TResult, [TKey]>,
  queryOptions?: QueryOptions<TResult, TError>
): QueryResult<TResult, TError>

```

`UseQueryRest` is not needed.

@tannerlinsley I'll be happy to open PR if you agree.

**Comment by MichaelDeBoey** [comment]
Is this still a valid problem now #767 is merged? 🤔

**Comment by boschni** [comment]
If I change the `react-query` version in the sandbox to `2.5.7-tsnext.1` then it seems to resolve the issue

**Comment by probablyup** [comment]
I'm still seeing this bug in 2.5.11

**Comment by tannerlinsley** [comment]
This is fixed in the new types (2.5.7.-tsnext.1). Soon to be released and set as the default types!

**Comment by probablyup** [comment]
Sorry I don't understand, there have been several releases since 2.5.7 and they don't seem to include these typings?
